[915] New back order logic after unpickable

parisiale opened the pull request.

parisiale committed:
[maint] Improve docstrings; style changes

parisiale committed:
[915] New back order logic after unpickable

With this commit we modify the logic for dealing with
unpickable items:
 - a back order is always created, on top of the investigation;
 - in case the unpickable item is available, such stock will
   be used for the very pick.

ws-mps commented on models/stock_picking_batch.py:
```diff
@@ -309,37 +303,55 @@ def unpickable_item(self, reason, product_id=None, location_id=None, package_nam
         remaining_pickings = self.picking_ids.filtered(
-            lambda x: x.state in ['assigned']
-        )
+            lambda x: x.state in ['assigned'])
```
What happens if we have the backorder in state 'waiting'? Probably we shouldn't be finishing the batch.
Add 'waiting' ?

parisiale replied:
Let me double check with the product fellows and get back to you.

ws-mps committed:
[915] Fix package stock.picking.batch unit tests

parisiale committed:
[915] Unbatch unpickable not 'assigned' picks

Here we ensure that, while processing unpickable picks,
we remove any non-assigned pick from the batch.

Previously, we were preventing the user to create a new
batch because a ongoing batch with a single 'waiting'
pick was assigned to him.

ws-mps committed:
[915] Fix unpickable item for products unit tests

ws-mps committed:
Merge branch 'story/915' of https://github.com/unipartdigital/udes_core into story/915

parisiale committed:
Merge remote-tracking branch 'origin/develop' into story/915

* origin/develop:
  [918] Cut long lines as per CR
  [918] Fix typo and use u_prev_picking_ids
  [918] Fix for canceled previous picks
  [FIX] Correcting field usage for dependencies
  [IMP] Adding Python Unit Tests for handle_partials functionality.
  [FIX] Removing circular dependency and unnecessary decorator.
  [IMP] Adding facility to block work on a picking until all its dependencies are completed.

parisiale committed:
Merge remote-tracking branch 'origin/story/915' into story/915

* origin/story/915:
  [915] Fix unpickable item for products unit tests

parisiale committed:
[915] Add new unit tests for new backorder logic

ws-mps commented on tests/test_picking_batch.py:
```diff
@@ -658,12 +704,59 @@ def test25_unpickable_item_product_ok_multiple_lines(self):
+                                      assign=True)
+
+        # check that only 1 unit of the quant is reserved
+        self.assertTrue(quant_apple_1.reserved_quantity == 1
```
As discussed, find the one we have reserved and check that the other is reserved after calling unpickable.

parisiale committed:
[915] Improve unpickable unit test

parisiale committed:
[915] More assertions for unpickable unit test
